docs/releases/4.1.2.txt

```diff
@@ -18,3 +18,6 @@ Bugfixes
 
 * Fixed a bug in Django 4.1 that caused an incorrect validation of
   ``CheckConstraint`` on ``NULL`` values (:ticket:`33996`).
+
+* Fixed classification of ``UniqueConstraint`` ``ValidationError``s with a
+  single field and condition (:ticket:`34007`).
```

Comment:
IMO it's not a bug but a cleanup so it doesn't qualify for a backport.

Reply:
Grrrr. 😜

I think the change from field to non-field error with the addition of the `condition` is pretty _surprising_ (shall we say). This is just introduced behaviour, so I don't think we should call it just a cleanup. 😬

(i.e. if this is worth correcting I think it's worth backporting as bug in a new feature.)

Reply:
Ah I see… I'll wait until classification of the ticket before updating any releases next time.

Reply:
> `condition=models.Q(color__isnull=True),`

I'd **guess** `condition`s referencing other fields like this would be rare. I grant that in that case it being a non-field error would make sense, but to switch all such for that possibility seems wrong to me. 

The initial kind of case from the ML discussion seems **much** more likely: 

```
class Test(Model):
    test = CharField(max_length=255)

    class Meta:
        constraints = [
            UniqueConstraint(fields=["test"], name="unique", condition=~Q(test="")),
        ]
```

... and that should clearly be a field error (I think).

Reply:
> Ah I see… I'll wait until classification of the ticket before updating any releases next time.

Let's keep the release note for now. The discussion is just beginning :boxing_glove::boxing_glove: :wink:

Reply:
oh I'd already taken it out 😅

nevermind I'll wait until resolution before touching anything else

Reply:
We should ask @charettes to decide 😀

Reply:
Yep agreed retrieving fields from Q could be too much here. I might take a look just out of curiosity though.

Reply:
@charettes I've had a go at your idea of retrieving fields from `Q` with a very simple property to collect children recursively, would something along these lines be suitable? You mentioned that a solution would need to be clever which makes me think there's more to it than I'm aware of 🤔 

I added a temp commit 36bcc9dbda71ad65fbb0777400144b12b2562327 to demo:

- Added a simple property `Q.fields` with a simple test
- Updated the validate() to check if the fields match the constraint. I personally don't think conditionally setting is the right way to go here – I agree with your comment on the ML about how the `len(fields) == 1 and code == "unique"` should probably move inside `UniqueConstraint.validate()` which means validation errors can have the code set unconditionally.
- Updated tests to show constraints being set as field & non-field errors based on whether the condition only refers to the field for which the constraint was declared.